File: app/src/main/java/com/attendance/tracker/ui/AttendanceViewModel.kt
package com.attendance.tracker.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.attendance.tracker.data.database.AttendanceDatabase
import com.attendance.tracker.data.model.AttendanceRecord
import com.attendance.tracker.data.model.AttendanceStatus
import com.attendance.tracker.data.model.ScheduleEntry
import com.attendance.tracker.data.model.Subject
import com.attendance.tracker.data.repository.AttendanceRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth

class AttendanceViewModel(application: Application) : AndroidViewModel(application) {
    private val database = AttendanceDatabase.getDatabase(application)
    private val repository = AttendanceRepository(
        database.subjectDao(),
        database.attendanceDao(),
        database.scheduleDao()
    )
    private val themeRepository = com.attendance.tracker.data.repository.ThemePreferenceRepository(
        database.themePreferenceDao()
    )

    // Undo/Redo Manager
    private val undoRedoManager = UndoRedoManager()
    
    private val _canUndo = MutableStateFlow(false)
    val canUndo: StateFlow<Boolean> = _canUndo.asStateFlow()
    
    private val _canRedo = MutableStateFlow(false)
    val canRedo: StateFlow<Boolean> = _canRedo.asStateFlow()

    // UI State
    val subjects: StateFlow<List<Subject>> = repository.actualSubjects
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())
    
    val allSubjectsIncludingFolders: StateFlow<List<Subject>> = repository.allSubjects
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())


    val scheduleEntries: StateFlow<List<ScheduleEntry>> = repository.allScheduleEntries
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _selectedDate = MutableStateFlow(LocalDate.now())
    val selectedDate: StateFlow<LocalDate> = _selectedDate.asStateFlow()

    private val _selectedMonth = MutableStateFlow(YearMonth.now())
    val selectedMonth: StateFlow<YearMonth> = _selectedMonth.asStateFlow()

    private val _todayAttendance = MutableStateFlow<Map<Long, AttendanceRecord>>(emptyMap())
    val todayAttendance: StateFlow<Map<Long, AttendanceRecord>> = _todayAttendance.asStateFlow()

    private val _attendanceRecords = MutableStateFlow<List<AttendanceRecord>>(emptyList())
    val attendanceRecords: StateFlow<List<AttendanceRecord>> = _attendanceRecords.asStateFlow()

    // Theme preferences
    val themePreference = themeRepository.themePreference
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    init {
        // Load today's attendance when ViewModel is created
        loadAttendanceForDate(LocalDate.now())
        // Initialize theme preferences
        viewModelScope.launch {
            themeRepository.initializeDefaultIfNeeded()
        }
    }

    fun loadAttendanceForDate(date: LocalDate) {
        viewModelScope.launch {
            repository.getAttendanceForDate(date).collect { records ->
                _todayAttendance.value = records.associateBy { it.subjectId }
            }
        }
    }

    fun loadAttendanceForMonth(yearMonth: YearMonth) {
        viewModelScope.launch {
            val startDate = yearMonth.atDay(1)
            val endDate = yearMonth.atEndOfMonth()
            repository.getAttendanceInRange(startDate, endDate).collect { records ->
                _attendanceRecords.value = records
            }
        }
    }

    fun setSelectedDate(date: LocalDate) {
        _selectedDate.value = date
        loadAttendanceForDate(date)
    }

    fun setSelectedMonth(yearMonth: YearMonth) {
        _selectedMonth.value = yearMonth
        loadAttendanceForMonth(yearMonth)
    }

    // Subject operations
    fun addSubject(name: String, requiredAttendance: Int = 75) {
        viewModelScope.launch {
            repository.insertSubject(
                Subject(name = name, requiredAttendance = requiredAttendance)
            )
        }
    }
    
    fun addSubjectFolder(name: String) {
        viewModelScope.launch {
            repository.insertSubject(
                Subject(name = name, isFolder = true)
            )
        }
    }
    
    fun addSubSubject(name: String, parentSubjectId: Long, requiredAttendance: Int = 75) {
        viewModelScope.launch {
            repository.insertSubject(
                Subject(
                    name = name, 
                    requiredAttendance = requiredAttendance,
                    parentSubjectId = parentSubjectId,
                    isFolder = false
                )
            )
        }
    }
    
    fun getSubSubjects(parentId: Long): Flow<List<Subject>> {
        return repository.getSubSubjects(parentId)
    }

    fun updateSubject(subject: Subject) {
        viewModelScope.launch {
            repository.updateSubject(subject)
        }
    }

    fun deleteSubject(subject: Subject) {
        viewModelScope.launch {
            repository.deleteSubject(subject)
        }
    }

    fun updateRequiredAttendance(subjectId: Long, required: Int) {
        viewModelScope.launch {
            repository.updateRequiredAttendance(subjectId, required)
        }
    }

    fun updateAttendanceCounts(subjectId: Long, present: Int, absent: Int) {
        viewModelScope.launch {
            repository.updateAttendanceCounts(subjectId, present, absent)
        }
    }

    // Attendance operations
    fun markAttendance(subjectId: Long, status: AttendanceStatus, date: LocalDate = LocalDate.now()) {
        viewModelScope.launch {
            // Get current state before marking
            val subject = repository.getSubjectById(subjectId)
            val oldRecord = repository.getAttendanceRecord(subjectId, date)
            
            if (subject != null) {
                // Record action for undo/redo
                val action = AttendanceAction(
                    subjectId = subjectId,
                    date = date,
                    oldStatus = oldRecord?.status,
                    newStatus = status,
                    oldPresentCount = subject.presentLectures,
                    oldAbsentCount = subject.absentLectures
                )
                undoRedoManager.recordAction(action)
                updateUndoRedoState()
                
                // Mark the new status
                when (status) {
                    AttendanceStatus.PRESENT -> repository.markPresent(subjectId, date)
                    AttendanceStatus.ABSENT -> repository.markAbsent(subjectId, date)
                    AttendanceStatus.NO_CLASS -> repository.markNoClass(subjectId, date)
                }
                loadAttendanceForDate(date)
            }
        }
    }
    
    fun undo() {
        viewModelScope.launch {
            val action = undoRedoManager.undo()
            if (action != null) {
                // Restore the old state
                val subject = repository.getSubjectById(action.subjectId)
                if (subject != null) {
                    // Restore attendance counts first
                    repository.updateAttendanceCounts(
                        action.subjectId,
                        action.oldPresentCount,
                        action.oldAbsentCount
                    )
                    
                    // Restore or delete the attendance record
                    if (action.oldStatus != null) {
                        // There was a previous status, restore it without modifying counts
                        repository.setAttendanceStatus(action.subjectId, action.date, action.oldStatus)
                    } else {
                        // No previous status, delete the record
                        repository.deleteAttendanceRecord(action.subjectId, action.date)
                    }
                    
                    loadAttendanceForDate(action.date)
                    updateUndoRedoState()
                }
            }
        }
    }
    
    fun redo() {
        viewModelScope.launch {
            val action = undoRedoManager.redo()
            if (action != null) {
                // Reapply the new status
                when (action.newStatus) {
                    AttendanceStatus.PRESENT -> repository.markPresent(action.subjectId, action.date)
                    AttendanceStatus.ABSENT -> repository.markAbsent(action.subjectId, action.date)
                    AttendanceStatus.NO_CLASS -> repository.markNoClass(action.subjectId, action.date)
                }
                loadAttendanceForDate(action.date)
                updateUndoRedoState()
            }
        }
    }
    
    private fun updateUndoRedoState() {
        _canUndo.value = undoRedoManager.canUndo
        _canRedo.value = undoRedoManager.canRedo
    }

    // Schedule operations
    fun addScheduleEntry(subjectId: Long, dayOfWeek: DayOfWeek) {
        viewModelScope.launch {
            repository.insertScheduleEntry(
                ScheduleEntry(subjectId = subjectId, dayOfWeek = dayOfWeek)
            )
        }
    }

    fun removeScheduleEntry(entry: ScheduleEntry) {
        viewModelScope.launch {
            repository.deleteScheduleEntry(entry)
        }
    }

    fun getScheduleForSubject(subjectId: Long): Flow<List<ScheduleEntry>> {
        return repository.getScheduleForSubject(subjectId)
    }

    fun getScheduleForDay(dayOfWeek: DayOfWeek): Flow<List<ScheduleEntry>> {
        return repository.getScheduleForDay(dayOfWeek)
    }

    // Theme operations
    fun updateThemeMode(themeMode: com.attendance.tracker.data.model.ThemeMode) {
        viewModelScope.launch {
            themeRepository.updateThemeMode(themeMode)
        }
    }

    fun updateCustomColors(primaryColor: Long?, secondaryColor: Long?) {
        viewModelScope.launch {
            themeRepository.updateCustomColors(primaryColor, secondaryColor)
        }
    }
}

File: iOS/AttendanceTracker/AttendanceTracker/ViewModels/AttendanceViewModel.swift
import Foundation
import SwiftData
import Combine

/// Action for undo/redo functionality
struct AttendanceAction {
    let subjectId: UUID
    let date: Date
    let oldStatus: AttendanceStatus?
    let newStatus: AttendanceStatus
    let oldPresentCount: Int
    let oldAbsentCount: Int
}

/// Manager for undo/redo functionality
class UndoRedoManager {
    private var undoStack: [AttendanceAction] = []
    private var redoStack: [AttendanceAction] = []
    
    var canUndo: Bool { !undoStack.isEmpty }
    var canRedo: Bool { !redoStack.isEmpty }
    
    func recordAction(_ action: AttendanceAction) {
        undoStack.append(action)
        redoStack.removeAll()
    }
    
    func undo() -> AttendanceAction? {
        guard let action = undoStack.popLast() else { return nil }
        redoStack.append(action)
        return action
    }
    
    func redo() -> AttendanceAction? {
        guard let action = redoStack.popLast() else { return nil }
        undoStack.append(action)
        return action
    }
}

/// Main ViewModel for the Attendance Tracker app
@MainActor
class AttendanceViewModel: ObservableObject {
    private let modelContext: ModelContext
    private let undoRedoManager = UndoRedoManager()
    
    @Published var subjects: [Subject] = []
    @Published var allSubjectsIncludingFolders: [Subject] = []
    @Published var scheduleEntries: [ScheduleEntry] = []
    @Published var selectedDate: Date = Date()
    @Published var selectedMonth: Date = Date()
    @Published var todayAttendance: [UUID: AttendanceStatus] = [:]
    @Published var attendanceRecords: [AttendanceRecord] = []
    @Published var canUndo: Bool = false
    @Published var canRedo: Bool = false
    
    init(modelContext: ModelContext) {
        self.modelContext = modelContext
        loadData()
        loadAttendanceForDate(Date())
    }
    
    // MARK: - Data Loading
    
    func loadData() {
        fetchSubjects()
        fetchScheduleEntries()
    }
    
    private func fetchSubjects() {
        do {
            let descriptor = FetchDescriptor<Subject>(sortBy: [SortDescriptor(\.name)])
            let allSubjects = try modelContext.fetch(descriptor)
            self.allSubjectsIncludingFolders = allSubjects
            self.subjects = allSubjects.filter { !$0.isFolder }
        } catch {
            print("Error fetching subjects: \(error)")
        }
    }
    
    private func fetchScheduleEntries() {
        do {
            let descriptor = FetchDescriptor<ScheduleEntry>()
            self.scheduleEntries = try modelContext.fetch(descriptor)
        } catch {
            print("Error fetching schedule entries: \(error)")
        }
    }
    
    func loadAttendanceForDate(_ date: Date) {
        do {
            let calendar = Calendar.current
            let startOfDay = calendar.startOfDay(for: date)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!
            
            let descriptor = FetchDescriptor<AttendanceRecord>(
                predicate: #Predicate { record in
                    record.date >= startOfDay && record.date < endOfDay
                }
            )
            let records = try modelContext.fetch(descriptor)
            self.todayAttendance = Dictionary(uniqueKeysWithValues: records.map { ($0.subjectId, $0.status) })
        } catch {
            print("Error loading attendance: \(error)")
        }
    }
    
    func loadAttendanceForMonth(_ date: Date) {
        do {
            let calendar = Calendar.current
            let startOfMonth = calendar.date(from: calendar.dateComponents([.year, .month], from: date))!
            let endOfMonth = calendar.date(byAdding: DateComponents(month: 1, day: -1), to: startOfMonth)!
            
            let descriptor = FetchDescriptor<AttendanceRecord>(
                predicate: #Predicate { record in
                    record.date >= startOfMonth && record.date <= endOfMonth
                }
            )
            self.attendanceRecords = try modelContext.fetch(descriptor)
        } catch {
            print("Error loading month attendance: \(error)")
        }
    }
    
    func setSelectedDate(_ date: Date) {
        selectedDate = date
        loadAttendanceForDate(date)
    }
    
    func setSelectedMonth(_ date: Date) {
        selectedMonth = date
        loadAttendanceForMonth(date)
    }
    
    // MARK: - Subject Operations
    
    func addSubject(name: String, requiredAttendance: Int = 75) {
        let subject = Subject(name: name, requiredAttendance: requiredAttendance)
        modelContext.insert(subject)
        saveContext()
        fetchSubjects()
    }
    
    func addSubjectFolder(name: String) {
        let folder = Subject(name: name, isFolder: true)
        modelContext.insert(folder)
        saveContext()
        fetchSubjects()
    }
    
    func addSubSubject(name: String, parentSubjectId: UUID, requiredAttendance: Int = 75) {
        let subject = Subject(
            name: name,
            requiredAttendance: requiredAttendance,
            parentSubjectId: parentSubjectId,
            isFolder: false
        )
        modelContext.insert(subject)
        saveContext()
        fetchSubjects()
    }
    
    func updateSubject(_ subject: Subject) {
        saveContext()
        fetchSubjects()
    }
    
    func deleteSubject(_ subject: Subject) {
        modelContext.delete(subject)
        saveContext()
        fetchSubjects()
    }
    
    func updateRequiredAttendance(subjectId: UUID, required: Int) {
        if let subject = allSubjectsIncludingFolders.first(where: { $0.id == subjectId }) {
            subject.requiredAttendance = required
            saveContext()
            fetchSubjects()
        }
    }
    
    func updateAttendanceCounts(subjectId: UUID, present: Int, absent: Int) {
        if let subject = allSubjectsIncludingFolders.first(where: { $0.id == subjectId }) {
            subject.presentLectures = present
            subject.absentLectures = absent
            subject.totalLectures = present + absent
            saveContext()
            fetchSubjects()
        }
    }
    
    // MARK: - Attendance Operations
    
    func markAttendance(subjectId: UUID, status: AttendanceStatus, date: Date = Date()) {
        guard let subject = allSubjectsIncludingFolders.first(where: { $0.id == subjectId }) else { return }
        
        // Get existing record for this date
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!
        
        do {
            let descriptor = FetchDescriptor<AttendanceRecord>(
                predicate: #Predicate { record in
                    record.subjectId == subjectId && record.date >= startOfDay && record.date < endOfDay
                }
            )
            let existingRecords = try modelContext.fetch(descriptor)
            let oldRecord = existingRecords.first
            
            // Record action for undo/redo
            let action = AttendanceAction(
                subjectId: subjectId,
                date: date,
                oldStatus: oldRecord?.status,
                newStatus: status,
                oldPresentCount: subject.presentLectures,
                oldAbsentCount: subject.absentLectures
            )
            undoRedoManager.recordAction(action)
            updateUndoRedoState()
            
            // Update or create attendance record
            if let existingRecord = oldRecord {
                existingRecord.status = status
            } else {
                let record = AttendanceRecord(subjectId: subjectId, date: date, status: status)
                modelContext.insert(record)
            }
            
            // Update subject counts
            switch status {
            case .present:
                subject.presentLectures += 1
                subject.totalLectures += 1
            case .absent:
                subject.absentLectures += 1
                subject.totalLectures += 1
            case .noClass:
                break
            }
            
            saveContext()
            loadAttendanceForDate(date)
            fetchSubjects()
        } catch {
            print("Error marking attendance: \(error)")
        }
    }
    
    func undo() {
        guard let action = undoRedoManager.undo() else { return }
        
        if let subject = allSubjectsIncludingFolders.first(where: { $0.id == action.subjectId }) {
            // Restore attendance counts
            subject.presentLectures = action.oldPresentCount
            subject.absentLectures = action.oldAbsentCount
            subject.totalLectures = action.oldPresentCount + action.oldAbsentCount
            
            // Restore or delete attendance record
            let calendar = Calendar.current
            let startOfDay = calendar.startOfDay(for: action.date)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!
            
            do {
                let descriptor = FetchDescriptor<AttendanceRecord>(
                    predicate: #Predicate { record in
                        record.subjectId == action.subjectId && record.date >= startOfDay && record.date < endOfDay
                    }
                )
                let records = try modelContext.fetch(descriptor)
                
                if let oldStatus = action.oldStatus, let record = records.first {
                    record.status = oldStatus
                } else if let record = records.first {
                    modelContext.delete(record)
                }
                
                saveContext()
                loadAttendanceForDate(action.date)
                fetchSubjects()
                updateUndoRedoState()
            } catch {
                print("Error in undo: \(error)")
            }
        }
    }
    
    func redo() {
        guard let action = undoRedoManager.redo() else { return }
        
        if let subject = allSubjectsIncludingFolders.first(where: { $0.id == action.subjectId }) {
            // Reapply the attendance
            let calendar = Calendar.current
            let startOfDay = calendar.startOfDay(for: action.date)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!
            
            do {
                let descriptor = FetchDescriptor<AttendanceRecord>(
                    predicate: #Predicate { record in
                        record.subjectId == action.subjectId && record.date >= startOfDay && record.date < endOfDay
                    }
                )
                let records = try modelContext.fetch(descriptor)
                
                if let record = records.first {
                    record.status = action.newStatus
                } else {
                    let record = AttendanceRecord(subjectId: action.subjectId, date: action.date, status: action.newStatus)
                    modelContext.insert(record)
                }
                
                // Update subject counts
                switch action.newStatus {
                case .present:
                    subject.presentLectures += 1
                    subject.totalLectures += 1
                case .absent:
                    subject.absentLectures += 1
                    subject.totalLectures += 1
                case .noClass:
                    break
                }
                
                saveContext()
                loadAttendanceForDate(action.date)
                fetchSubjects()
                updateUndoRedoState()
            } catch {
                print("Error in redo: \(error)")
            }
        }
    }
    
    private func updateUndoRedoState() {
        canUndo = undoRedoManager.canUndo
        canRedo = undoRedoManager.canRedo
    }
    
    // MARK: - Schedule Operations
    
    func addScheduleEntry(subjectId: UUID, dayOfWeek: DayOfWeek) {
        let entry = ScheduleEntry(subjectId: subjectId, dayOfWeek: dayOfWeek)
        modelContext.insert(entry)
        saveContext()
        fetchScheduleEntries()
    }
    
    func removeScheduleEntry(_ entry: ScheduleEntry) {
        modelContext.delete(entry)
        saveContext()
        fetchScheduleEntries()
    }
    
    func getScheduleForSubject(_ subjectId: UUID) -> [ScheduleEntry] {
        scheduleEntries.filter { $0.subjectId == subjectId }
    }
    
    func getScheduleForDay(_ dayOfWeek: DayOfWeek) -> [ScheduleEntry] {
        scheduleEntries.filter { $0.dayOfWeek == dayOfWeek }
    }
    
    // MARK: - Helper
    
    private func saveContext() {
        do {
            try modelContext.save()
        } catch {
            print("Error saving context: \(error)")
        }
    }
}
